pkg/api/registered/registered.go

```diff
@@ -25,71 +26,126 @@ import (
-		{Group: "", Version: "v1"},
-		{Group: "extensions", Version: "v1beta1"},
-		{Group: "componentconfig", Version: "v1alpha1"},
-	}
+	// envRequestedVersions represents the versions requested via the
+	// KUBE_API_VERSIONS environment variable. The install package of each group
+	// checks this list before add their versions to the latest package and
+	// Scheme.
+	envRequestedVersions = map[unversioned.GroupVersion]struct{}{}
+)
 
-	// Env var KUBE_API_VERSIONS is a comma separated list of API versions that should be registered in the scheme.
-	// The versions should be in the order of most preferred to the least.
-	userRequestedVersions := os.Getenv("KUBE_API_VERSIONS")
-	if len(userRequestedVersions) != 0 {
-		// reset the supported versions
-		supportedVersions = []unversioned.GroupVersion{}
-		for _, version := range strings.Split(userRequestedVersions, ",") {
+func init() {
+	// Env var KUBE_API_VERSIONS is a comma separated list of API versions that
+	// should be registered in the scheme.
+	kubeAPIVersions := os.Getenv("KUBE_API_VERSIONS")
+	if len(kubeAPIVersions) != 0 {
+		for _, version := range strings.Split(kubeAPIVersions, ",") {
 			gv, err := unversioned.ParseGroupVersion(version)
 			if err != nil {
-				glog.Fatalf("invalid api version: %s in KUBE_API_VERSIONS: %s. List of valid API versions: %v",
-					version, os.Getenv("KUBE_API_VERSIONS"), validGroupVersions)
+				glog.Fatalf("invalid api version: %s in KUBE_API_VERSIONS: %s.",
+					version, os.Getenv("KUBE_API_VERSIONS"))
 			}
+			envRequestedVersions[gv] = struct{}{}
+		}
+	}
+}
 
-			// Verify that the version is valid.
-			valid, ok := validGroupVersions[gv]
-			if !ok || !valid {
-				// Not a valid API version.
-				glog.Fatalf("invalid api version: %s in KUBE_API_VERSIONS: %s. List of valid API versions: %v",
-					version, os.Getenv("KUBE_API_VERSIONS"), validGroupVersions)
-			}
+// RegisterVersions add the versions the registeredVersions.
+func RegisterVersions(versions ...unversioned.GroupVersion) {
+	for _, v := range versions {
+		registeredVersions[v] = struct{}{}
+	}
+}
 
-			supportedVersions = append(supportedVersions, gv)
+// EnableVersions add the versions to the enabledVersions.
+func EnableVersions(versions ...unversioned.GroupVersion) error {
+	var unregisteredVersions []unversioned.GroupVersion
+	for _, v := range versions {
+		if _, found := registeredVersions[v]; !found {
+			unregisteredVersions = append(unregisteredVersions, v)
 		}
+		enabledVersions[v] = struct{}{}
+	}
+	if len(unregisteredVersions) != 0 {
+		fmt.Errorf("Please register versions before enabling them: %v", unregisteredVersions)
 	}
+	return nil
+}
 
-	RegisteredGroupVersions = supportedVersions
+// IsEnvReuqestedVersion returns if the version is requested via the
+// KUBE_API_VERSIONS environment variable. If the environment variable is empty,
+// then it always returns true.
+func IsEnvReuqestedVersion(v unversioned.GroupVersion) bool {
```

Comment:
> But I'm afraid that will lead people to think the versions listed in the Env are the excluded versions. I think explain the "no env set" case in the comment is clearer.

Ok, how about `IsAllowedVersion`?  Also, the name is currently typoed: Requested